File: hash/HashMap.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define INITIAL_CAPACITY 16
#define LOAD_FACTOR_THRESHOLD 0.75

typedef struct Entry {
    char *key;
    int value;
    struct Entry *next;
} Entry;

typedef struct HashMap {
    Entry **buckets;
    int capacity;
    int size;
} HashMap;

// Create a new entry
Entry* createEntry(const char *key, int value) {
    Entry *entry = (Entry*)malloc(sizeof(Entry));
    entry->key = (char*)malloc(strlen(key) + 1);
    strcpy(entry->key, key);
    entry->value = value;
    entry->next = NULL;
    return entry;
}

// Hash function using djb2 algorithm
unsigned int hash(const char *key, int capacity) {
    unsigned long hashValue = 5381;
    int c;
    
    while ((c = *key++))
        hashValue = ((hashValue << 5) + hashValue) + c;
    
    return hashValue % capacity;
}

// Initialize hash map
HashMap* createHashMap() {
    HashMap *map = (HashMap*)malloc(sizeof(HashMap));
    map->capacity = INITIAL_CAPACITY;
    map->size = 0;
    map->buckets = (Entry**)calloc(map->capacity, sizeof(Entry*));
    return map;
}

// Get load factor
double getLoadFactor(HashMap *map) {
    return (double)map->size / map->capacity;
}

// Resize and rehash the map
void resize(HashMap *map) {
    int oldCapacity = map->capacity;
    Entry **oldBuckets = map->buckets;
    
    map->capacity *= 2;
    map->buckets = (Entry**)calloc(map->capacity, sizeof(Entry*));
    map->size = 0;
    
    // Rehash all entries
    for (int i = 0; i < oldCapacity; i++) {
        Entry *entry = oldBuckets[i];
        while (entry != NULL) {
            Entry *next = entry->next;
            
            // Reinsert into new buckets
            unsigned int index = hash(entry->key, map->capacity);
            entry->next = map->buckets[index];
            map->buckets[index] = entry;
            map->size++;
            
            entry = next;
        }
    }
    
    free(oldBuckets);
}

// Put a key-value pair into the map
void put(HashMap *map, const char *key, int value) {
    // Check if resize is needed
    if (getLoadFactor(map) >= LOAD_FACTOR_THRESHOLD) {
        resize(map);
    }
    
    unsigned int index = hash(key, map->capacity);
    Entry *entry = map->buckets[index];
    
    // Check if key already exists
    while (entry != NULL) {
        if (strcmp(entry->key, key) == 0) {
            entry->value = value;
            return;
        }
        entry = entry->next;
    }
    
    // Insert new entry at the beginning
    Entry *newEntry = createEntry(key, value);
    newEntry->next = map->buckets[index];
    map->buckets[index] = newEntry;
    map->size++;
}

// Get value for a key
int get(HashMap *map, const char *key, bool *found) {
    unsigned int index = hash(key, map->capacity);
    Entry *entry = map->buckets[index];
    
    while (entry != NULL) {
        if (strcmp(entry->key, key) == 0) {
            *found = true;
            return entry->value;
        }
        entry = entry->next;
    }
    
    *found = false;
    return -1;
}

// Check if key exists
bool containsKey(HashMap *map, const char *key) {
    unsigned int index = hash(key, map->capacity);
    Entry *entry = map->buckets[index];
    
    while (entry != NULL) {
        if (strcmp(entry->key, key) == 0)
            return true;
        entry = entry->next;
    }
    
    return false;
}

// Remove a key-value pair
bool removeKey(HashMap *map, const char *key) {
    unsigned int index = hash(key, map->capacity);
    Entry *entry = map->buckets[index];
    Entry *prev = NULL;
    
    while (entry != NULL) {
        if (strcmp(entry->key, key) == 0) {
            if (prev == NULL)
                map->buckets[index] = entry->next;
            else
                prev->next = entry->next;
            
            free(entry->key);
            free(entry);
            map->size--;
            return true;
        }
        prev = entry;
        entry = entry->next;
    }
    
    return false;
}

// Get all keys
char** getKeys(HashMap *map, int *numKeys) {
    char **keys = (char**)malloc(map->size * sizeof(char*));
    int index = 0;
    
    for (int i = 0; i < map->capacity; i++) {
        Entry *entry = map->buckets[i];
        while (entry != NULL) {
            keys[index] = (char*)malloc(strlen(entry->key) + 1);
            strcpy(keys[index], entry->key);
            index++;
            entry = entry->next;
        }
    }
    
    *numKeys = index;
    return keys;
}

// Get all values
int* getValues(HashMap *map, int *numValues) {
    int *values = (int*)malloc(map->size * sizeof(int));
    int index = 0;
    
    for (int i = 0; i < map->capacity; i++) {
        Entry *entry = map->buckets[i];
        while (entry != NULL) {
            values[index++] = entry->value;
            entry = entry->next;
        }
    }
    
    *numValues = index;
    return values;
}

// Clear all entries
void clear(HashMap *map) {
    for (int i = 0; i < map->capacity; i++) {
        Entry *entry = map->buckets[i];
        while (entry != NULL) {
            Entry *next = entry->next;
            free(entry->key);
            free(entry);
            entry = next;
        }
        map->buckets[i] = NULL;
    }
    map->size = 0;
}

// Check if map is empty
bool isEmpty(HashMap *map) {
    return map->size == 0;
}

// Get size of map
int getSize(HashMap *map) {
    return map->size;
}

// Print the hash map
void printHashMap(HashMap *map) {
    printf("HashMap (size: %d, capacity: %d, load factor: %.2f):\n", 
           map->size, map->capacity, getLoadFactor(map));
    
    for (int i = 0; i < map->capacity; i++) {
        if (map->buckets[i] != NULL) {
            printf("Bucket %d: ", i);
            Entry *entry = map->buckets[i];
            while (entry != NULL) {
                printf("[%s: %d] ", entry->key, entry->value);
                entry = entry->next;
            }
            printf("\n");
        }
    }
}

// Free the hash map
void freeHashMap(HashMap *map) {
    clear(map);
    free(map->buckets);
    free(map);
}

// Example usage
int main() {
    HashMap *map = createHashMap();
    
    // Insert key-value pairs
    printf("Inserting elements...\n");
    put(map, "apple", 100);
    put(map, "banana", 200);
    put(map, "orange", 300);
    put(map, "grape", 400);
    put(map, "mango", 500);
    
    printHashMap(map);
    
    // Get values
    printf("\nGetting values:\n");
    bool found;
    int value = get(map, "banana", &found);
    printf("banana: %s (value: %d)\n", found ? "found" : "not found", value);
    
    value = get(map, "cherry", &found);
    printf("cherry: %s\n", found ? "found" : "not found");
    
    // Check if key exists
    printf("\nContains key 'apple': %s\n", containsKey(map, "apple") ? "yes" : "no");
    printf("Contains key 'cherry': %s\n", containsKey(map, "cherry") ? "yes" : "no");
    
    // Update value
    printf("\nUpdating 'apple' to 150...\n");
    put(map, "apple", 150);
    value = get(map, "apple", &found);
    printf("apple: %d\n", value);
    
    // Get all keys
    printf("\nAll keys: ");
    int numKeys;
    char **keys = getKeys(map, &numKeys);
    for (int i = 0; i < numKeys; i++) {
        printf("%s ", keys[i]);
        free(keys[i]);
    }
    free(keys);
    printf("\n");
    
    // Get all values
    printf("All values: ");
    int numValues;
    int *values = getValues(map, &numValues);
    for (int i = 0; i < numValues; i++) {
        printf("%d ", values[i]);
    }
    free(values);
    printf("\n");
    
    // Remove a key
    printf("\nRemoving 'banana'...\n");
    removeKey(map, "banana");
    printf("Size after removal: %d\n", getSize(map));
    
    // Test resizing by adding many elements
    printf("\nAdding more elements to trigger resize...\n");
    for (int i = 0; i < 20; i++) {
        char key[20];
        sprintf(key, "key%d", i);
        put(map, key, i * 10);
    }
    
    printHashMap(map);
    
    // Clear the map
    printf("\nClearing map...\n");
    clear(map);
    printf("Size after clear: %d\n", getSize(map));
    printf("Is empty: %s\n", isEmpty(map) ? "yes" : "no");
    
    // Free memory
    freeHashMap(map);
    
    return 0;
}

File: hash/HashMap.java
class Entry {
    String key;
    int value;
    Entry next;
    
    public Entry(String key, int value) {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

public class HashMap {
    private static final int INITIAL_CAPACITY = 16;
    private static final double LOAD_FACTOR_THRESHOLD = 0.75;
    
    private Entry[] buckets;
    private int capacity;
    private int size;
    
    public HashMap() {
        this.capacity = INITIAL_CAPACITY;
        this.size = 0;
        this.buckets = new Entry[capacity];
    }
    
    // Hash function
    private int hash(String key) {
        int hashValue = key.hashCode();
        return Math.abs(hashValue) % capacity;
    }
    
    // Get load factor
    private double getLoadFactor() {
        return (double) size / capacity;
    }
    
    // Resize and rehash the map
    private void resize() {
        int oldCapacity = capacity;
        Entry[] oldBuckets = buckets;
        
        capacity *= 2;
        buckets = new Entry[capacity];
        size = 0;
        
        // Rehash all entries
        for (int i = 0; i < oldCapacity; i++) {
            Entry entry = oldBuckets[i];
            while (entry != null) {
                put(entry.key, entry.value);
                entry = entry.next;
            }
        }
    }
    
    // Put a key-value pair into the map
    public void put(String key, int value) {
        // Check if resize is needed
        if (getLoadFactor() >= LOAD_FACTOR_THRESHOLD) {
            resize();
        }
        
        int index = hash(key);
        Entry entry = buckets[index];
        
        // Check if key already exists
        while (entry != null) {
            if (entry.key.equals(key)) {
                entry.value = value;
                return;
            }
            entry = entry.next;
        }
        
        // Insert new entry at the beginning
        Entry newEntry = new Entry(key, value);
        newEntry.next = buckets[index];
        buckets[index] = newEntry;
        size++;
    }
    
    // Get value for a key
    public Integer get(String key) {
        int index = hash(key);
        Entry entry = buckets[index];
        
        while (entry != null) {
            if (entry.key.equals(key)) {
                return entry.value;
            }
            entry = entry.next;
        }
        
        return null;
    }
    
    // Check if key exists
    public boolean containsKey(String key) {
        int index = hash(key);
        Entry entry = buckets[index];
        
        while (entry != null) {
            if (entry.key.equals(key))
                return true;
            entry = entry.next;
        }
        
        return false;
    }
    
    // Remove a key-value pair
    public boolean remove(String key) {
        int index = hash(key);
        Entry entry = buckets[index];
        Entry prev = null;
        
        while (entry != null) {
            if (entry.key.equals(key)) {
                if (prev == null)
                    buckets[index] = entry.next;
                else
                    prev.next = entry.next;
                
                size--;
                return true;
            }
            prev = entry;
            entry = entry.next;
        }
        
        return false;
    }
    
    // Get all keys
    public String[] getKeys() {
        String[] keys = new String[size];
        int index = 0;
        
        for (int i = 0; i < capacity; i++) {
            Entry entry = buckets[i];
            while (entry != null) {
                keys[index++] = entry.key;
                entry = entry.next;
            }
        }
        
        return keys;
    }
    
    // Get all values
    public int[] getValues() {
        int[] values = new int[size];
        int index = 0;
        
        for (int i = 0; i < capacity; i++) {
            Entry entry = buckets[i];
            while (entry != null) {
                values[index++] = entry.value;
                entry = entry.next;
            }
        }
        
        return values;
    }
    
    // Get all entries
    public Entry[] getEntries() {
        Entry[] entries = new Entry[size];
        int index = 0;
        
        for (int i = 0; i < capacity; i++) {
            Entry entry = buckets[i];
            while (entry != null) {
                entries[index++] = entry;
                entry = entry.next;
            }
        }
        
        return entries;
    }
    
    // Clear all entries
    public void clear() {
        for (int i = 0; i < capacity; i++) {
            buckets[i] = null;
        }
        size = 0;
    }
    
    // Check if map is empty
    public boolean isEmpty() {
        return size == 0;
    }
    
    // Get size of map
    public int size() {
        return size;
    }
    
    // Get capacity
    public int capacity() {
        return capacity;
    }
    
    // Print the hash map
    public void printHashMap() {
        System.out.printf("HashMap (size: %d, capacity: %d, load factor: %.2f):%n", 
                         size, capacity, getLoadFactor());
        
        for (int i = 0; i < capacity; i++) {
            if (buckets[i] != null) {
                System.out.print("Bucket " + i + ": ");
                Entry entry = buckets[i];
                while (entry != null) {
                    System.out.print("[" + entry.key + ": " + entry.value + "] ");
                    entry = entry.next;
                }
                System.out.println();
            }
        }
    }
    
    // Print all key-value pairs
    public void printAll() {
        System.out.println("All entries:");
        for (Entry entry : getEntries()) {
            System.out.println(entry.key + " -> " + entry.value);
        }
    }
    
    // Example usage
    public static void main(String[] args) {
        HashMap map = new HashMap();
        
        // Insert key-value pairs
        System.out.println("Inserting elements...");
        map.put("apple", 100);
        map.put("banana", 200);
        map.put("orange", 300);
        map.put("grape", 400);
        map.put("mango", 500);
        
        map.printHashMap();
        
        // Get values
        System.out.println("\nGetting values:");
        Integer value = map.get("banana");
        System.out.println("banana: " + (value != null ? "found (value: " + value + ")" : "not found"));
        
        value = map.get("cherry");
        System.out.println("cherry: " + (value != null ? "found" : "not found"));
        
        // Check if key exists
        System.out.println("\nContains key 'apple': " + map.containsKey("apple"));
        System.out.println("Contains key 'cherry': " + map.containsKey("cherry"));
        
        // Update value
        System.out.println("\nUpdating 'apple' to 150...");
        map.put("apple", 150);
        value = map.get("apple");
        System.out.println("apple: " + value);
        
        // Get all keys
        System.out.print("\nAll keys: ");
        for (String key : map.getKeys()) {
            System.out.print(key + " ");
        }
        System.out.println();
        
        // Get all values
        System.out.print("All values: ");
        for (int val : map.getValues()) {
            System.out.print(val + " ");
        }
        System.out.println();
        
        // Remove a key
        System.out.println("\nRemoving 'banana'...");
        map.remove("banana");
        System.out.println("Size after removal: " + map.size());
        
        // Test resizing by adding many elements
        System.out.println("\nAdding more elements to trigger resize...");
        for (int i = 0; i < 20; i++) {
            map.put("key" + i, i * 10);
        }
        
        map.printHashMap();
        
        System.out.println("\nAll entries after resize:");
        map.printAll();
        
        // Clear the map
        System.out.println("\nClearing map...");
        map.clear();
        System.out.println("Size after clear: " + map.size());
        System.out.println("Is empty: " + map.isEmpty());
    }
}
